File: program/rectangle.cpp
//------------------------------------------------------------------------------
// rectangle.cpp - содержит процедуры для работы с прямоугольником
//------------------------------------------------------------------------------

#define _CRT_SECURE_NO_WARNINGS
#include "stdio.h"
#include "rectangle.h"

//------------------------------------------------------------------------------
// Ввод параметров прямоугольника из файла
void In(rectangle& r, FILE* file) {
    fscanf(file, "%d", &r.x_l_t);
    fscanf(file, "%d", &r.y_l_t);
    fscanf(file, "%d", &r.x_r_b);
    fscanf(file, "%d", &r.y_r_b);
}

// Случайный ввод параметров прямоугольника
void InRnd(rectangle& r) {
    r.x_l_t = Random();
    r.y_l_t = Random();
    do {
        r.x_r_b = Random();
    } while (r.x_r_b == r.x_l_t);
    do {
        r.y_r_b = Random();
    } while (r.y_r_b == r.y_l_t);
}

//------------------------------------------------------------------------------
// Вывод параметров прямоугольника в форматируемый поток
void Out(rectangle& r, FILE* file) {
    fprintf(file, "It is Rectangle. Left top point: x = %d, y = %d. Right bottom point: x = %d, y = %d. Perimeter = %f. Color: ",
        r.x_l_t, r.y_l_t, r.x_r_b, r.y_r_b, Perimeter(r));
}

//------------------------------------------------------------------------------
// Вычисление периметра прямоугольника
double Perimeter(rectangle& r) {
    int x = std::abs(r.x_l_t - r.x_r_b);
    int y = std::abs(r.y_l_t - r.y_r_b);
    return 2.0 * (x + y);
}

File: hw3/program/rectangle.py
import color
import random
from shape import *
# ----------------------------------------------


class Rectangle(Shape):
    def __init__(self):
        self.x_l_t = 0
        self.y_l_t = 0
        self.x_r_b = 0
        self.y_r_b = 0
        self.color = 0

    def get_figure(self, data, i):
        self.color = int(data[i])
        self.x_l_t = int(data[i+1])
        self.y_l_t = int(data[i+2])
        self.x_r_b = int(data[i+3])
        self.y_r_b = int(data[i+4])

    def get_random_figure(self):
        self.x_l_t = random.randint(0, 21)
        self.y_l_t = random.randint(0, 21)
        self.x_r_b = random.randint(0, 21)
        while self.x_r_b == self.x_l_t:
            self.x_r_b = random.randint(0, 21)
        self.y_r_b = random.randint(0, 21)
        while self.y_r_b == self.y_l_t:
            self.y_r_b = random.randint(0, 21)

    def perimeter(self):
        x = abs(self.x_l_t - self.x_r_b)
        y = abs(self.y_l_t - self.y_r_b)
        return 2.0 * (x + y)

    def write(self, output):
        output.write("Rectangle: x (top left) = {}  y (top left) = {}, "
                     "x (bottom right) = {}  y (bottom right) = {}. "
                     "Perimeter = {}. Color is {}"
                     .format(self.x_l_t, self.y_l_t, self.x_r_b, self.y_r_b,
                             self.perimeter(), color.get_color(self.color)))
